### PR TITLE
zstandard compression for parameters and results

### PR DESCRIPTION
**Pull request by guzzijones**
# why
mitigate (as much as possible) 'document too large' mongo exceptions.  16 mb is the limit for mongodb document size.

# done
1.  zstandard compression for parameter and results
2.  remove embedded liveaction doc in executions db and replace with liveaction id string. This change will  potentially save 1/2 the time to write the liveaction to the database as the liveaction won't be written 2 times.
3.  added traceback info to exception handling for action execution engine and workflow engine.  This should make it easier to debug for someone when the workflow fails for some reason.  invalid orjson or unterminated yaql strings cause these exceptions in my experience.
4. This will require a data migration for the `executionDB` model for the `liveaction` paramter as it is now a string as opposed to an embedded document
5. I added a config option to turn off zstandard compression for parameters and results in the execution db and the liveaction db

### CODE REVIEW

File: st2common/st2common/models/db/execution.py [hunk 1]
```diff
@@ -61,29 +51,39 @@ class ActionExecutionDB(stormbase.StormFoundationDB):
...
     )
+    delay = me.IntField(min_value=0)
+
+    # diff from liveaction
+    runner = stormbase.EscapedDictField(required=True)
+    trigger = stormbase.EscapedDictField()
+    trigger_type = stormbase.EscapedDictField()
+    trigger_instance = stormbase.EscapedDictField()
+    rule = stormbase.EscapedDictField()
+    result_size = me.IntField(default=0, help_text="Serialized result size in bytes")
```
Comment (guzzijones) [inline st2common/st2common/models/db/execution.py:73]: todo: This is currently being set to the zstandard compressed size not the serialized size.  I need to instead set this to the  bytes  size before compression

File: st2common/st2common/models/db/execution.py [hunk 2]
```diff
@@ -61,29 +51,39 @@ class ActionExecutionDB(stormbase.StormFoundationDB):
...
     web_url = me.StringField(required=False)
 
+    # liveaction id
+    liveaction = me.StringField()
+
     meta = {
         "indexes": [
             {"fields": ["rule.ref"]},
             {"fields": ["action.ref"]},
-            {"fields": ["liveaction.id"]},
```
Comment (guzzijones) [inline st2common/st2common/models/db/execution.py:86]: todo: add liveaction as an indexed field

File: st2common/st2common/models/db/execution.py [hunk 3]
```diff
@@ -39,16 +38,7 @@
 class ActionExecutionDB(stormbase.StormFoundationDB):
     RESOURCE_TYPE = ResourceType.EXECUTION
     UID_FIELDS = ["id"]
-
```
Comment (guzzijones) [inline st2common/st2common/models/db/execution.py:42]: liveaction was changed from an embedded dict to a string.  all other changes here are cosmetic

File: st2common/st2common/models/db/liveaction.py [hunk 4]
```diff
@@ -54,11 +55,7 @@ class LiveActionDB(stormbase.StormFoundationDB):
     action = me.StringField(
         required=True, help_text="Reference to the action that has to be executed."
     )
-    action_is_workflow = me.BooleanField(
```
Comment (guzzijones) [inline st2common/st2common/models/db/liveaction.py:57]: cosmetic changes here

File: st2common/st2common/openapi.yaml.j2 [hunk 5]
```diff
@@ -4927,7 +4927,7 @@ definitions:
       runner:
         $ref: '#/definitions/RunnerType'
       liveaction:
-        $ref: '#/definitions/LiveAction'
+        type: string
```
Comment (guzzijones) [inline st2common/st2common/openapi.yaml.j2:4930]: double check this endpoint doesn't return the liveaction.  I think there was a unit test that I forced the liveaction to return embedded.  not sure if this is it.

File: st2common/st2common/services/policies.py [hunk 6]
```diff
@@ -58,9 +61,11 @@ def apply_pre_run_policies(lv_ac_db):
             LOG.info(message % (policy_db.ref, policy_db.policy_type, str(lv_ac_db.id)))
             lv_ac_db = driver.apply_before(lv_ac_db)
         except:
-            message = 'An exception occurred while applying policy "%s" (%s) for liveaction "%s".'
+            _, ex, tb = sys.exc_info()
```
Comment (guzzijones) [inline st2common/st2common/services/policies.py:64]: traceback was added here for additional debugging info for the user.

File: st2common/tests/unit/migrations/test_v35_migrate_db_dict_field_values.py [hunk 7]
```diff
@@ -80,10 +80,20 @@ def test_migrate_executions(self):
 
         class ActionExecutionDB_OldFieldType(ActionExecutionDB):
             result = stormbase.EscapedDynamicField(default={})
+            liveaction = stormbase.EscapedDictField(required=True)
+            parameters = stormbase.EscapedDynamicField(default={})
 
         class LiveActionDB_OldFieldType(LiveActionDB):
             result = stormbase.EscapedDynamicField(default={})
 
+        #todo(aj) need ActionExecutionDB_NewFieldType to be used for update
```
Comment (guzzijones) [inline st2common/tests/unit/migrations/test_v35_migrate_db_dict_field_values.py:89]: todo: need to hard code the state of this model so this unit test works again

File: st2common/tests/unit/test_db_execution.py [hunk 8]
```diff
@@ -188,19 +191,6 @@ def test_execution_inquiry_secrets(self):
             "supersecretvalue",
         )
 
-    def test_execution_inquiry_response_action(self):
```
Comment (guzzijones) [inline st2common/tests/unit/test_db_execution.py:191]: todo: why did I remove this test?

File: st2common/tests/unit/test_db_fields.py [hunk 9]
```diff
@@ -147,78 +148,17 @@ def test_parse_field_value(self):
         self.assertEqual(result, {"c": "d"})
 
 
-class JSONDictFieldTestCaseWithHeader(unittest2.TestCase):
```
Comment (guzzijones) [inline st2common/tests/unit/test_db_fields.py:150]: old zstandard header testing removed here.

File: st2common/st2common/openapi.yaml.j2 (hunk 5 above)
Comment (guzzijones) [reply st2common/st2common/openapi.yaml.j2:4930]: it was the aliasexecution endpoint I had to embedd the liveaction not this endpoint.  this endpoint returns the execution id now

File: st2common/st2common/models/db/execution.py (hunk 2 above)
Comment (guzzijones) [reply st2common/st2common/models/db/execution.py:86]: set liveaction to an index field

File: st2common/tests/unit/migrations/test_v35_migrate_db_dict_field_values.py (hunk 7 above)
Comment (guzzijones) [reply st2common/tests/unit/migrations/test_v35_migrate_db_dict_field_values.py]: Is this test worth keeping around?  It assumes that the current model structure is the final model structure ever.

File: st2common/st2common/models/db/execution.py (hunk 1 above)
Comment (guzzijones) [reply st2common/st2common/models/db/execution.py:74]: This should be resolved now.

**Comment by guzzijones** [comment]
# Benchmark results

TLDR. 
zstandard is many times faster for single field large json
zstandard is slightly slower for 8mb json

The cost here is minimal from a speed perspective

```
$ /st2/st2common/benchmarks/micro$ pytest test_mongo_field_types.py --benchmark-columns=mean,rounds

--------------------------------- benchmark 'live_action_read': 12 tests --------------------------------
Name (time in ms)                                                                  Mean            Rounds
---------------------------------------------------------------------------------------------------------
test_read_large_execution[old_json_dict_field-tiny_1]                            2.3494 (1.0)         387
test_read_large_execution[json_dict_field-tiny_1]                                2.3697 (1.01)        391
test_read_large_execution[old_json_dict_field-json_61kb]                         3.2883 (1.40)        254
test_read_large_execution[json_dict_field-json_61kb]                             3.3988 (1.45)        269
test_read_large_execution[json_dict_field-json_4mb_single_large_field]          12.3281 (5.25)         78
test_read_large_execution[old_json_dict_field-json_647kb]                       13.6826 (5.82)         66
test_read_large_execution[json_dict_field-json_647kb]                           15.9034 (6.77)         67
test_read_large_execution[old_json_dict_field-json_4mb_single_large_field]      17.2653 (7.35)         56
test_read_large_execution[old_json_dict_field-json_4mb]                        114.0750 (48.55)        11
test_read_large_execution[json_dict_field-json_4mb]                            115.1644 (49.02)        11
test_read_large_execution[json_dict_field-json_8mb]                            248.3191 (105.69)        5
test_read_large_execution[old_json_dict_field-json_8mb]                        246.7735 (105.03)        5
---------------------------------------------------------------------------------------------------------

--------------------------------- benchmark 'live_action_save': 12 tests --------------------------------
Name (time in ms)                                                                  Mean            Rounds
---------------------------------------------------------------------------------------------------------
test_save_large_execution[old_json_dict_field-tiny_1]                            6.5457 (1.0)          25
test_save_large_execution[json_dict_field-tiny_1]                                6.7954 (1.04)        134
test_save_large_execution[old_json_dict_field-json_61kb]                        11.6969 (1.79)         78
test_save_large_execution[json_dict_field-json_61kb]                            13.1218 (2.00)         75
test_save_large_execution[json_dict_field-json_4mb_single_large_field]          18.3983 (2.81)         54
test_save_large_execution[old_json_dict_field-json_4mb_single_large_field]      30.7534 (4.70)         28
test_save_large_execution[old_json_dict_field-json_647kb]                       73.0526 (11.16)        14
test_save_large_execution[json_dict_field-json_647kb]                           81.4653 (12.45)        13
test_save_large_execution[old_json_dict_field-json_4mb]                        381.4303 (58.27)         5
test_save_large_execution[json_dict_field-json_4mb]                            385.8807 (58.95)         5
test_save_large_execution[old_json_dict_field-json_8mb]                        737.6227 (112.69)        5
test_save_large_execution[json_dict_field-json_8mb]                            756.7961 (115.62)        5
---------------------------------------------------------------------------------------------------------

-------------------------- benchmark 'live_action_save_multiple_fields': 10 tests -------------------------
Name (time in ms)                                                                    Mean            Rounds
-----------------------------------------------------------------------------------------------------------
test_save_multiple_fields[old_json_dict_field-tiny_1]                              6.6645 (1.0)         113
test_save_multiple_fields[json_dict_field-tiny_1]                                  6.7131 (1.01)        123
test_save_multiple_fields[old_json_dict_field-json_61kb]                          22.1928 (3.33)         47
test_save_multiple_fields[json_dict_field-json_61kb]                              24.7535 (3.71)         41
test_save_multiple_fields[json_dict_field-json_4mb_single_large_field]            37.2357 (5.59)         25
test_save_multiple_fields[old_json_dict_field-json_4mb_single_large_field]       119.9479 (18.00)        10
test_save_multiple_fields[old_json_dict_field-json_647kb]                        198.9707 (29.86)         5
test_save_multiple_fields[json_dict_field-json_647kb]                            228.3664 (34.27)         5
test_save_multiple_fields[old_json_dict_field-json_4mb]                        1,144.1683 (171.68)        5
test_save_multiple_fields[json_dict_field-json_4mb]                            1,139.4643 (170.98)        5
-----------------------------------------------------------------------------------------------------------

```

**Comment by guzzijones** [comment]
This will require a data migration for the `executionDB` model for the `liveaction` paramter as it is now a string as opposed to an embedded document

**Comment by guzzijones** [comment]
# data migration tested

```
$ ./st2-migrate-liveaction-executiondb  --config-file ../../../../conf/st2.dev.conf
StackStorm v3.5 database field data migration script

Will migrate objects with creation date between 2023-06-06 13:01:20 UTC and 2023-07-06 13:01:20 UTC.

You are strongly recommended to create database backup before proceeding.

Depending on the number of the objects in the database, migration may take multiple hours or more. You are recommended to start the script in a screen session, tmux or similar.

To proceed with the migration, press enter and to cancel it, press CTRL+C.


Migrating affected database objects between 2023-06-06 13:01:20 and 2023-07-06 13:01:20

Migrating execution objects
Will migrate 10 ActionExecutionDB objects

[1/10] Migrating ActionExecutionDB with id 64a6ba4fdce54d96db28f4ae
ActionExecutionDB with id 64a6ba4fdce54d96db28f4ae has been migrated
[2/10] Migrating ActionExecutionDB with id 64a6ba40dce54d96db28f4a2
ActionExecutionDB with id 64a6ba40dce54d96db28f4a2 has been migrated
[3/10] Migrating ActionExecutionDB with id 64a6ba3ddce54d96db28f49c
ActionExecutionDB with id 64a6ba3ddce54d96db28f49c has been migrated
[4/10] Migrating ActionExecutionDB with id 64a6ba4edce54d96db28f4ab
ActionExecutionDB with id 64a6ba4edce54d96db28f4ab has been migrated
[5/10] Migrating ActionExecutionDB with id 64a6ba50dce54d96db28f4b7
ActionExecutionDB with id 64a6ba50dce54d96db28f4b7 has been migrated
[6/10] Migrating ActionExecutionDB with id 64a6ba3fdce54d96db28f49f
ActionExecutionDB with id 64a6ba3fdce54d96db28f49f has been migrated
[7/10] Migrating ActionExecutionDB with id 64a6ba4fdce54d96db28f4b1
ActionExecutionDB with id 64a6ba4fdce54d96db28f4b1 has been migrated
[8/10] Migrating ActionExecutionDB with id 64a6ba4edce54d96db28f4a8
ActionExecutionDB with id 64a6ba4edce54d96db28f4a8 has been migrated
[9/10] Migrating ActionExecutionDB with id 64a6ba41dce54d96db28f4a5
ActionExecutionDB with id 64a6ba41dce54d96db28f4a5 has been migrated
[10/10] Migrating ActionExecutionDB with id 64a6ba50dce54d96db28f4b4
ActionExecutionDB with id 64a6ba50dce54d96db28f4b4 has been migrated
SUCCESS: All database objects migrated successfully (duration: 0 seconds).

```

**Comment by guzzijones** [comment]
1. add a field for compression of parameters somewhere.

File: st2common/st2common/models/api/base.py [hunk 10]
```diff
@@ -94,32 +101,36 @@ def _from_model(cls, model, mask_secrets=False):
...
                 raw_values[field_name] = doc.pop(field_name)
-
-        # TODO (Tomaz): In general we really shouldn't need to call unescape chars on the whole doc,
-        # but just on the EscapedDict and EscapedDynamicField fields - doing it on the whole doc
-        # level is slow and not necessary!
-        doc = util_mongodb.unescape_chars(doc)
-
-        # Now add the JSON string field value which shouldn't be escaped back.
-        # We don't JSON parse the field value here because that happens inside the model specific
-        # "from_model()" method where we also parse and convert all the other field values.
```
Comment (cognifloyd) [inline st2common/st2common/models/api/base.py:114]: This note should be available in the convert_raw method (either in the docstring or as a comment).

File: st2common/st2common/models/db/execution.py [hunk 11]
```diff
@@ -128,31 +128,6 @@ def mask_secrets(self, value):
...
             parameters=result.get("parameters", {}), secret_parameters=secret_parameters
         )
-
-        if "parameters" in liveaction:
-            liveaction["parameters"] = mask_secret_parameters(
-                parameters=liveaction["parameters"], secret_parameters=secret_parameters
-            )
-
-            if liveaction.get("action", "") == "st2.inquiry.respond":
-                # Special case to mask parameters for `st2.inquiry.respond` action
```
Comment (cognifloyd) [inline st2common/st2common/models/db/execution.py:138]: Where does this handling happen now?

**Review by cognifloyd — CHANGES_REQUESTED** [review]

File: st2common/tests/unit/test_db_execution.py (hunk 8 above)
Comment (guzzijones) [reply st2common/tests/unit/test_db_execution.py:191]: the live action is no longer embedded in the execution returned here so there is nothing to mask.

File: st2api/tests/unit/controllers/v1/test_alias_execution.py [hunk 12]
```diff
@@ -159,6 +159,7 @@ def test_execution_secret_parameter(self, request):
         self.assertEqual(post_resp.status_int, 201)
         expected_parameters = {"param1": "value1", "param4": SUPER_SECRET_PARAMETER}
         self.assertEqual(request.call_args[0][0].parameters, expected_parameters)
+        # above working
```
[the remark this answers is not in the document]
Comment (guzzijones) [reply st2api/tests/unit/controllers/v1/test_alias_execution.py:162]: I will remove it.  looks like some notes i forgot about

File: st2common/benchmarks/micro/test_mongo_field_types.py [hunk 13]
```diff
@@ -71,46 +115,31 @@ class LiveActionDB_EscapedDynamicField(LiveActionDB):
...
+    field3 = stormbase.EscapedDynamicField(default={})
 
 
-class LiveActionDB_JSONFieldWithHeader(LiveActionDB):
-    result = JSONDictField(default={}, use_header=True, compression_algorithm="none")
+# 3. Old Approach which uses no compression where value is stored as serialized JSON string / blob
+class LiveActionDB_OLDJSONField(LiveActionDB):
+    result = OldJSONDictField(default={}, use_header=False)
 
-    field1 = JSONDictField(default={}, use_header=True, compression_algorithm="none")
```
[the remark this answers is not in the document]
Comment (guzzijones) [reply st2common/benchmarks/micro/test_mongo_field_types.py:95]: I am using a header character again so we can support other compression languages.

File: st2common/tests/unit/test_db_execution.py (hunk 8 above)
Comment (guzzijones) [reply st2common/tests/unit/test_db_execution.py:191]: I suppose we can move this test over to the live action model to be sure.

File: st2common/st2common/models/db/execution.py (hunk 2 above)
Comment (guzzijones) [reply st2common/st2common/models/db/execution.py:86]: > When adding new fields or new indexes, MongoEngine will take care of creating them (unless auto_create_index is disabled)

So yes mongo will create the index, but it isn't clear that the old history with the full document will be valid.

Comment (guzzijones) [reply st2common/st2common/models/db/execution.py:86]: I suppose changing the field to `liveaction_id` could ensure we don't have issues with the migration of this field.

File: st2common/st2common/models/db/execution.py (hunk 11 above)
Comment (cognifloyd) [reply st2common/st2common/models/db/execution.py:138]: That makes sense. Please move the inquiry masking into the live action model since that is what the API will be retrieving.

File: st2common/st2common/fields.py [hunk 14]
```diff
@@ -358,25 +361,14 @@ class JSONDictField(BinaryField):
...
-
-    If header is not used, value is stored as a serialized JSON string of the input dictionary.
     """
 
     def __init__(self, *args, **kwargs):
-        # True if we should use field header which is more future proof approach and also allows
-        # us to support optional per-field compression, etc.
-        # This option is only exposed so we can benchmark different approaches and how much overhead
-        # using a header adds.
-        self.use_header = kwargs.pop("use_header", False)
```
[the remark this answers is not in the document]
Comment (guzzijones) [reply st2common/st2common/fields.py:377]: I ended using a header by default now.

**Comment by guzzijones** [comment]
I changed to `liveaction_id`.

File: st2api/tests/unit/controllers/v1/test_alias_execution.py [hunk 15]
```diff
@@ -322,15 +322,8 @@ def test_match_and_execute_list_action_param_str_cast_to_list(self):
         self.assertEqual(resp.status_int, 201)
 
         result = resp.json["results"][0]
-        live_action = result["execution"]["liveaction"]
         action_alias = result["actionalias"]
-
         self.assertEqual(resp.status_int, 201)
-        self.assertTrue(isinstance(live_action["parameters"]["array_param"], list))
```
Comment (guzzijones) [inline st2api/tests/unit/controllers/v1/test_alias_execution.py:329]: LiveAction is no longer an embedded document in the actionexecution table.   If users are using that output then it will not longer exist.  Do we need to make a new controller for liveaction so it can be retrieved?

**Review by cognifloyd — COMMENTED** [review]
For the migration, I wonder if there's a raw query we can use that delegates the migration to mongo. Something like  `{"$set": {"liveaction_id": doc.liveaction.id}}`

File: st2common/st2common/models/api/execution.py [hunk 16]
```diff
@@ -60,7 +60,7 @@ class ActionExecutionAPI(BaseAPI):
             "rule": RuleAPI.schema,
             "action": REQUIRED_ATTR_SCHEMAS["action"],
             "runner": REQUIRED_ATTR_SCHEMAS["runner"],
-            "liveaction": REQUIRED_ATTR_SCHEMAS["liveaction"],
+            "liveaction_id": {"type": "string", "required": True},
```
Comment (cognifloyd) [inline st2common/st2common/models/api/execution.py:63]: The API does not have to match the database. So, we can leave this as liveaction in the API model but construct it from both execution and liveaction db models. That would make this backwards compatible because the public API would not be changing.

**Comment by guzzijones** [comment]
Since `liveaction` field of the ActionExecutionDB was required before I made `liveaction_id` also required.  This should help ensure the liveaction actually exists.

File: st2api/st2api/controllers/v1/execution_views.py [hunk 17]
```diff
@@ -31,7 +31,7 @@
 SUPPORTED_FILTERS = {
     "action": "action.ref",
     "status": "status",
-    "liveaction": "liveaction.id",
+    "liveaction_id": "liveaction_id",
```
[the remark this answers is not in the document]
Comment (guzzijones) [reply st2api/st2api/controllers/v1/execution_views.py:34]: I added `liveaction` back with a field value of `liveaction_id`. so it will function like before.

File: st2common/st2common/models/api/execution.py (hunk 16 above)
Comment (guzzijones) [reply st2common/st2common/models/api/execution.py:65]: I added the embedded liveaction back to the database.

File: st2api/tests/unit/controllers/v1/test_executions_filters.py [hunk 18]
```diff
@@ -89,8 +95,8 @@ def assign_parent(child):
             data["id"] = obj_id
             data["start_timestamp"] = isotime.format(timestamp, offset=False)
             data["end_timestamp"] = isotime.format(timestamp, offset=False)
-            data["status"] = data["liveaction"]["status"]
-            data["result"] = data["liveaction"]["result"]
+            data["status"] = data["status"]
+            data["result"] = data["result"]
```
[the remark this answers is not in the document]
Comment (guzzijones) [reply st2api/tests/unit/controllers/v1/test_executions_filters.py]: deleted.

**Comment by guzzijones** [comment]
also, what is up with the circle-ci failures?  is there anything we can do to fix that?

File: st2common/tests/unit/test_db_execution.py (hunk 8 above)
Comment (cognifloyd) [inline st2common/tests/unit/test_db_execution.py:191]: Would you please move this to `test_db_liveaction.py`?

File: st2common/st2common/fields.py [hunk 19]
```diff
@@ -406,55 +370,27 @@ def parse_field_value(self, value: Optional[Union[bytes, dict]]) -> dict:
...
-            data = zstandard.ZstdDecompressor().decompress(data)
+        data = value
+        try:
+            uncompression_header = value[0:1]
+            uncompression_method = MAP_UNCOMPRESS.get(uncompression_header, False)
+            if uncompression_method:
+                data = uncompression_method(value[1:])
+            # skip if already a byte string and not compressed
+        except zstandard.ZstdError:
+            pass
```
[the remark this answers is not in the document]
Comment (guzzijones) [reply st2common/st2common/fields.py:388]: sounds good. will do.

File: st2common/st2common/fields.py [hunk 20]
```diff
@@ -474,21 +410,14 @@ def default(obj):
...
 
-        if not self.use_header:
-            return orjson.dumps(value, default=default)
-
         data = orjson.dumps(value, default=default)
+        parameter_result_compression = cfg.CONF.database.parameter_result_compression
+        compression_method = MAP_COMPRESS.get(parameter_result_compression, False)
+        # none is not mapped at all so has no compression method
+        if compress and compression_method:
+            data = compression_method(data)
```
[the remark this answers is not in the document]
Comment (guzzijones) [reply st2common/st2common/fields.py:418]: sounds good. will do.

File: st2common/tests/unit/test_db_execution.py (hunk 8 above)
Comment (guzzijones) [reply st2common/tests/unit/test_db_execution.py:191]: sounds good. will do.

Comment (guzzijones) [reply st2common/tests/unit/test_db_execution.py:191]: done

File: st2common/st2common/fields.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (guzzijones) [reply st2common/st2common/fields.py]: done

File: st2common/st2common/fields.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (guzzijones) [reply st2common/st2common/fields.py]: done

**Review by cognifloyd — APPROVED** [review]
Minor nit.

I don't know why Circle CI is falling. I'm guessing something we really on is no longer supported or was upgraded.

**Comment by guzzijones** [comment]
@cognifloyd The circle ci is fixed.  rabbitmq needs priv=true and redis needed to be pinned. [pr](https://github.com/StackStorm/st2-packages/pull/727). can you merge?

**Comment by Kami** [comment]
I think zstandard compression is a good change in case numbers across various types of datasets confirm it's indeed worth it / good trade off of CPU cycles used for compressing / decompressing those field values.

We decided to not enable compression by default as part of #4846 to keep things simple and reduce the scope of those changes. I'm fine with enabling the compression by default now, in case it numbers still back this up.

As far as this change goes:

> remove embedded liveaction doc in executions db and replace with liveaction id string. This change will potentially save 1/2 the time to write the liveaction to the database as the liveaction won't be written 2 times.

This one I'm not too sure about it yet and I need to dig in and think about it some more.

I think it would be good if we split this PR into two - one for field compression changes and one for field value / foreign key change.

In the end, both changes are somewhat related, but they have vastly different implications - one was already benchmarked in the past, it's backward / forward compatible and the other one is not, it could have large implications and also requires a potentially expensive and breaking migration (which is not backward compatible).

**Review by Kami — DISMISSED** [review]
Thanks for your contribution.

I think we are on the right track, but I added some comments on splitting this PR into two and potentially adding more micro benchmarks, etc.

**Comment by guzzijones** [comment]
FWIW we have forked over to this branch  along with removing the [deep copies in orquesta ](https://github.com/StackStorm/orquesta/pull/256).   We are seeing huge speed improvements.  Also, no more Mongo document too large.

**Comment by guzzijones** [comment]
I rebased the changes.  Now that we completed the mongo, python3.9, 3.10 migration code can we look at this again?

**Comment by guzzijones** [comment]
Are we any closer to a release so we can eventually merge this?

**Comment by guzzijones** [comment]
are we ready to merge this?

**Comment by guzzijones** [comment]
@nzlosh @skiedude this one is now ready. i will get the orquesta changes next.

**Review by nzlosh — APPROVED** [review]

**Review by skiedude — APPROVED** [review]
